### PR TITLE
[Android/iOS] TabIndex sets focus order for TalkBack traversal

### LINKED ISSUE #2789: Improved A11y Support - TabIndex
## Description
As an app developer, I currently am unable to specify an expected focus/tab order for elements in my Xamarin.Forms app that will be used by a screen reader and/or keyboard navigation to improve the usability and accessibility of the app.  When this spec is complete, I can explicitly set the `TabIndex` on any focusable element to have it read to the app user and/or prepare it for input in a particular order.

### Acceptance criteria
* Solution must enable app developer to set an integer value on `Element.TabIndex` with the expectation that focus will cycle through each element in ascending order. 
* Conflicts on `TabIndex` will be resolved by the declaration order. 
* If no value is set in `TabIndex`, the default behavior will be used. 
* If the app developer attempts to set a focus order using a custom renderer, the values set by the custom renderer should take precedence.

#### Per the UWP specification:
* UI elements with `TabIndex` equal to 0 are added to the tab order based on declaration order in XAML or child collections.
* UI elements with `TabIndex` greater than 0 are added to the tab order based on the `TabIndex` value.
* UI elements with `TabIndex` less than 0 are added to the tab order and appear before any zero value.

## API changes
Add: `int Element.TabIndex { get; set; }`

### Description of changes
- UWP: https://docs.microsoft.com/en-us/windows/uwp/accessibility/keyboard-accessibility
- iOS: https://developer.apple.com/documentation/objectivec/nsobject/1615147-accessibilityelements#declarations 
- Android: https://developer.android.com/training/keyboard-input/navigation.html#Tab 
- macOS: https://github.com/xamarin/Xamarin.Forms/pull/1060 (convert from platform specific)

### PR DESCRIPTION
**Pull request by samhouts**
### Description of Change ###

Setting the `TabIndex` of a `VisualElement` now affects the order in which TalkBack/VoiceOver will read the accessible elements on the screen.

### Issues Resolved ### 

- fixes #2789

#### Type Changed: Xamarin.Forms.TabIndexExtensions

Added methods:

```csharp
public static VisualElement GetFirstElementByTabIndex (System.Collections.Generic.IDictionary<System.Int32,System.Collections.Generic.List<VisualElement>> tabIndexes);
public static VisualElement GetFirstNonLayoutTabStop (System.Collections.Generic.IDictionary<System.Int32,System.Collections.Generic.List<VisualElement>> tabIndexes);
```

#### Type Changed: Xamarin.Forms.Platform.Android.PageRenderer

Added method:

```csharp
protected override void OnLayout (bool changed, int l, int t, int r, int b);
```

#### New Type: Xamarin.Forms.Platform.iOS.PageContainer

```csharp
public class PageContainer : UIKit.UIView, ObjCRuntime.INativeObject, System.IDisposable, UIKit.IUIAccessibilityContainer {
  // constructors
  public PageContainer ();
  public PageContainer (AccessibleUIViewController parent);
  // methods
  public void ClearAccessibilityElements ();
  protected override void Dispose (bool disposing);
}
```

### Platforms Affected ### 

- Core/XAML (all platforms)
- Android
- iOS

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### Testing Procedure ###
1. Turn on TalkBack on your Android device or VoiceOver on your iOS device.
2. Open Control Gallery.
3. Navigate to Accessibility TabIndex test page.
4. Swipe right to traverse through the elements. Each element is labeled with its TabIndex (first) and its declaration order (second). You should see it select elements in order of TabIndex first and then DeclarationOrder for all elements with a conflicting TabIndex.

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

**Comment by kingces95** [comment]
Save for weekend.

**Comment by PureWeen** [comment]
@samhouts  all the android tests failed on the same two. Maybe it just needs to be rebased?

**Comment by samhouts** [comment]
build --uitests

**Comment by kingces95** [comment]
When I launch the A11y TabIndex page the, "This text should be read when the page loads", is not read. I can click on it and then it'll be read.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		<Grid.RowDefinitions>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="Auto"></RowDefinition>
+			<RowDefinition Height="Auto"></RowDefinition>
+		</Grid.RowDefinitions>
+
+		<BoxView Color="Blue"></BoxView>
+		<Label Text="This text should be read when the page loads."  TextColor="White" />
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml]: Yeah, flaw in the test page, not in the code, fortunately.

File: Xamarin.Forms.Core.UnitTests/TabIndexTests.cs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+					new Label { TabIndex = 2 },
+				}
+			};
+
+			var page = new ContentPage { Content = target };
+
+			var tabIndexes = target.GetTabIndexesOnParentPage(out int _);
+
+			//StackLayout is technically the first element with TabIndex 0.
+			Assert.AreEqual(target, tabIndexes[0][0]);
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core.UnitTests/TabIndexTests.cs]: The first group `tabIndexes[0]` returns all elements with `TabIndex == 0`, of which there can be many. Then you grab the first element in that list, which should be sorted by declaration order.

File: Xamarin.Forms.sln [hunk 3]
```diff
@@ -172,7 +172,7 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Xamarin.Forms.Xaml.Design",
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "XFCorePostProcessor.Tasks", "XFCorePostProcessor.Tasks\XFCorePostProcessor.Tasks.csproj", "{5BBF4A3F-4AD1-47FD-B250-05EA793F939D}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Xamarin.Forms.Platform.Tizen (Forwarders)", "Stubs\Xamarin.Forms.Platform.Tizen\Xamarin.Forms.Platform.Tizen (Forwarders).csproj", "{39B3457F-01D8-43D0-8E84-D8C4F73CF48E}"
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.sln]: Yeah, should revert it

File: Xamarin.Forms.Core.UnitTests/TabIndexTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kingces95) [reply Xamarin.Forms.Core.UnitTests/TabIndexTests.cs]: Ok. And not that it matters, but why? 

Why not return the elements in the order they'd be traversed by swiping as outlined in the reproduction instructions. At that point our API simply returns a struct which could implement IEnumerable. If the caller wants to group 'em, have at it but why pre-group 'em? 

I could see pre-grouping if the API were internal and was simply a factoring out of logic. But for a public API, why not go simple and return an enumerator?

File: Xamarin.Forms.Core/TabIndexExtensions.cs [hunk 4]
```diff
@@ -80,5 +83,24 @@ public static VisualElement FindNextElement(this VisualElement element, bool for
 				}
 			}
 		}
+
+		public static VisualElement GetFirstNonLayoutTabStop(IDictionary<int, List<VisualElement>> tabIndexes)
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core/TabIndexExtensions.cs]: Moved to Internals namespace

File: Xamarin.Forms.Core/TabIndexExtensions.cs [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
 
 namespace Xamarin.Forms
 {
 	public static class TabIndexExtensions
 	{
-		public static IDictionary<int, List<VisualElement>> GetTabIndexesOnParentPage(this VisualElement element, out int countChildrensWithTabStopWithoutThis)
+		public static IDictionary<int, List<VisualElement>> GetTabIndexesOnParentPage(this VisualElement element, out int countChildrenWithTabStopWithoutThis)
 		{
-			countChildrensWithTabStopWithoutThis = 0;
+			countChildrenWithTabStopWithoutThis = 0;
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core/TabIndexExtensions.cs]: yeah, just removed the double plural.

File: Xamarin.Forms.Core/TabIndexExtensions.cs [hunk 6]
```diff
@@ -1,8 +1,7 @@
 using System.Collections.Generic;
-using Xamarin.Forms.Internals;
 using System.Linq;
 
-namespace Xamarin.Forms
+namespace Xamarin.Forms.Internals
```
Comment (kingces95) [inline Xamarin.Forms.Core/TabIndexExtensions.cs]: Happy day!

File: Xamarin.Forms.Core.UnitTests/TabIndexTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core.UnitTests/TabIndexTests.cs]: Basically, I'm basing my work off of the work that @paymicro did for the TabIndexes, which didn't plan on me reusing the `FindNextElement` method, so it didn't care about the usability of the method signature. Do you think we'd see a performance improvement with your suggestion?

File: Xamarin.Forms.Core.UnitTests/TabIndexTests.cs [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+			var page = new ContentPage { Content = stackLayout };
+
+			var tabIndexes = stackLayout.GetTabIndexesOnParentPage(out int maxAttempts);
+
+			var first = TabIndexExtensions.GetFirstNonLayoutTabStop(tabIndexes);
+
+			Assert.AreEqual(target, first);
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core.UnitTests/TabIndexTests.cs]: There aren't, but there should be. `maxAttempts` is irrelevant for my PR here, so I confess that I ignored it.

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Core/TabIndexExtensions.cs (hunk 6 above)
Comment (samhouts) [reply Xamarin.Forms.Core/TabIndexExtensions.cs]: 💃

File: Xamarin.Forms.Core/TabIndexExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Core/TabIndexExtensions.cs]: Also, yeah, the `.Min` is gross, but I determined that any code path I had to follow given the types I had at my disposal would be less performant. Fortunately, this code path is only called when the screen readers are on, which I believe kill perf more than an enumerator will.

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Core.UnitTests/TabIndexTests.cs [hunk 8]
```diff
@@ -0,0 +1,278 @@
+using NUnit.Framework;
+using System.Linq;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms.Core.UnitTests
+{
+	[TestFixture]
+	public class TabIndexTests : BaseTestFixture
+	{
+
```
Comment (paymicro) [inline Xamarin.Forms.Core.UnitTests/TabIndexTests.cs]: Can add test without explicit tabindexes? To make sure all the items are taken in the correct order.
```suggestion
		[Test]
		public void FindNextElement_Forward_Without_Tabindexes()
		{
			var stackLayout = new StackLayout();

			var labels = new Label[5];
			for (int i = 0; i < labels.Length; i++)
			{
				labels[i] = new Label();
				stackLayout.Children.Add(labels[i]);
			}

			var page = new ContentPage { Content = stackLayout };

			var tabIndexes = stackLayout.GetTabIndexesOnParentPage(out int __);

			for (int i = 0; i < labels.Length; i++)
			{
				int _ = labels[i].TabIndex;
				var found = labels[i].FindNextElement(true, tabIndexes, ref _);
				if (labels.Length != i + 1)
					Assert.AreEqual(labels[i + 1], found);
				else
					Assert.AreEqual(stackLayout, found);
			}
		}

```

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Platform.iOS/Renderers/PageContainer.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+			{
+				// lazy-loading this list so that the expensive call to GetAccessibilityElements only happens when VoiceOver is on.
+				if (_accessibilityElements == null)
+				{
+					_accessibilityElements = _parent.GetAccessibilityElements();
+					if (_accessibilityElements == null)
+					{
+						NSObject defaultElements = AccessibilityContainer.GetAccessibilityElements();
+						if (defaultElements != null)
+							_accessibilityElements = NSArray.ArrayFromHandle<NSObject>(AccessibilityContainer.GetAccessibilityElements().Handle).ToList();
```
Comment (paymicro) [inline Xamarin.Forms.Platform.iOS/Renderers/PageContainer.cs]:
```suggestion
							_accessibilityElements = NSArray.ArrayFromHandle<NSObject>(defaultElements.Handle).ToList();
```

Comment (samhouts) [reply Xamarin.Forms.Platform.iOS/Renderers/PageContainer.cs]: whoops, thanks!!

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		<Grid.RowDefinitions>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="50"></RowDefinition>
+			<RowDefinition Height="Auto"></RowDefinition>
+			<RowDefinition Height="Auto"></RowDefinition>
+		</Grid.RowDefinitions>
+
+		<BoxView Color="Blue"></BoxView>
+		<Label Text="This text should be read when the page loads." TabIndex="-2" TextColor="White" />
```
Comment (PureWeen) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml]: Update instructions to match how this runs in our control gallery.

From my testing in the control gallery when navigating to this page nothing is read to the user. Is that correct? 
First swipe says "Main Page main page help text" 
Second swipe reads back button
third swipe reads this text

Testing on Android S9

Comment (PureWeen) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml]: The behavior swiping back and forth is a bit interesting on Android
- onload nothing plays
- swipe right "Main Page"
- swipe right "Gallery Button"
- swipe left "Main Page" 
- swipe left - *some sort of thunk sound?*
- swipe right "Gallery Button"

Not sure if it's related to this PR or not but figured I'd mention it

**Comment by PureWeen** [comment]
@samhouts  it looks like a lot of the UI Tests are still failing

File: Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs [hunk 11]
```diff
@@ -115,5 +123,57 @@ void UpdateBackgroundImage(Page view)
 			if (!string.IsNullOrEmpty(view.BackgroundImage))
 				this.SetBackground(Context.GetDrawable(view.BackgroundImage));
 		}
+
+		void IOrderedTraversalController.UpdateTraversalOrder()
+		{
+			// traversal order wasn't added until API 22
+			if ((int)Build.VERSION.SdkInt < 22)
```
Comment (PureWeen) [inline Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]: Could we use the Compat library?
https://developer.android.com/reference/android/support/v4/view/accessibility/AccessibilityNodeInfoCompat

```C#
global::Android.Support.V4.View.Accessibility.AccessibilityNodeInfoCompat.Wrap(control.CreateAccessibilityNodeInfo())
						.SetTraversalAfter(thisControl);
```

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml (hunk 10 above)
Comment (PureWeen) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml]: Update instructions to match how this runs in our control gallery.

From my testing in the control gallery when navigating to this page nothing is read to the user. Is that correct? 
First swipe says "Main Page main page help text" 
Second swipe reads back button
third swipe reads this text

Testing on Android S9


The behavior swiping back and forth is a bit interesting on Android
- onload nothing plays
- swipe right "Main Page"
- swipe right "Gallery Button"
- swipe left "Main Page" 
- swipe left - *some sort of thunk sound?*
- swipe right "Gallery Button"

Not sure if it's related to this PR or not but figured I'd mention it

File: Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs (hunk 11 above)
Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]: Could we use the Compat library?
https://developer.android.com/reference/android/support/v4/view/accessibility/AccessibilityNodeInfoCompat

```C#
global::Android.Support.V4.View.Accessibility.AccessibilityNodeInfoCompat.Wrap(control.CreateAccessibilityNodeInfo())
						.SetTraversalAfter(thisControl);
```

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml (hunk 10 above)
Comment (PureWeen) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/A11yTabIndex.xaml]: Update instructions to match how this runs in our control gallery.

From my testing in the control gallery when navigating to this page nothing is read to the user. Is that correct?
First swipe says "Main Page main page help text"
Second swipe reads back button
third swipe reads this text

Testing on Android S9



The behavior swiping back and forth is a bit interesting on Android

- onload nothing plays
- swipe right "Main Page"
- swipe right "Gallery Button"
- swipe left "Main Page"
- swipe left - some sort of thunk sound?
- swipe right "Gallery Button"

Not sure if it's related to this PR or not but figured I'd mention it

File: Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs (hunk 11 above)
Comment (PureWeen) [inline Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]: Could we use the Compat library?
https://developer.android.com/reference/android/support/v4/view/accessibility/AccessibilityNodeInfoCompat

```C#
global::Android.Support.V4.View.Accessibility.AccessibilityNodeInfoCompat.Wrap(control.CreateAccessibilityNodeInfo())
						.SetTraversalAfter(thisControl);
```

Comment (samhouts) [reply Xamarin.Forms.Platform.Android/Renderers/PageRenderer.cs]: Tried, but it seems to do nothing.

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
build --uitests

**Comment by PureWeen** [comment]
The 0 2 label seems to activate differently for iOS vs Android

### Case 1
on android it goes: 0 2, -2,  -1 (-1), 0 3
on iOS it goes: -2,  -1 (-1), 0 2, 0 3

### Case 2
on iOS if I swipe to the end it stops on "4 14" and I can't go any further
on android if I swipe to the end it stops on "Skip 3 6", I swipe and it bings at me like an error, then I swipe again and it selects the page

**Comment by samhouts** [comment]
Yeah, looks like Android is busted again. DNM! Thanks

**Comment by samhouts** [comment]
@PureWeen K, so, it should read the elements in the proper order now on Android. However, it will incorrectly read the initial text and the Skip button. This is due to #4270, and it should be considered outside the scope of this PR. 

I can't reproduce Case 2. Expectations are that iOS should just loop back to the beginning; Android will incorrectly read the master page contents first (#4220), then BONG that you're at the end of the elements; wait a second, and then it will allow you to start from the beginning again.

**Pull request merged**